From 6a340ba0e60a12796e61a81e6669e6577bcc1c21 Mon Sep 17 00:00:00 2001
From: Florent Tomasin <florent.tomasin@nxp.com>
Date: Wed, 12 Sep 2018 11:22:37 +0300
Subject: [PATCH 4602/5242] MLK-18170-1: ASoC: imx-wm8960: use API jack
 functions

commit  50eeeac1695e8e3c93121e5c2d6551c16122aaab from
https://source.codeaurora.org/external/imx/linux-imx.git

Fix the build when IMX_WM8960 machine driver is compiled as module:
CONFIG_SND_SOC_IMX_WM8960=m

Signed-off-by: Florent Tomasin <florent.tomasin@nxp.com>
Signed-off-by: Viorel Suman <viorel.suman@nxp.com>
Reviewed-by: Shengjiu Wang <shengjiu.wang@nxp.com>
(cherry picked from commit 233f1db5c861034acb29ade544d05604d08ef040)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-wm8960.c |   20 ++++++++++++--------
 1 file changed, 12 insertions(+), 8 deletions(-)

diff --git a/sound/soc/fsl/imx-wm8960.c b/sound/soc/fsl/imx-wm8960.c
index 82f87ba..93f8cf6 100644
--- a/sound/soc/fsl/imx-wm8960.c
+++ b/sound/soc/fsl/imx-wm8960.c
@@ -44,7 +44,7 @@ struct imx_priv {
 	enum of_gpio_flags hp_active_low;
 	enum of_gpio_flags mic_active_low;
 	bool is_headset_jack;
-	struct snd_kcontrol *headphone_kctl;
+	struct snd_jack jack_data;
 	struct platform_device *pdev;
 	struct platform_device *asrc_pdev;
 	struct snd_card *snd_card;
@@ -92,7 +92,7 @@ static int hp_jack_status_check(void *data)
 			snd_soc_dapm_disable_pin(dapm, "Main MIC");
 		}
 		ret = imx_hp_jack_gpio.report;
-		snd_kctl_jack_report(priv->snd_card, priv->headphone_kctl, 1);
+		snd_jack_report(jack->jack, 1);
 	} else {
 		snd_soc_dapm_enable_pin(dapm, "Ext Spk");
 		if (priv->is_headset_jack) {
@@ -100,7 +100,7 @@ static int hp_jack_status_check(void *data)
 			snd_soc_dapm_enable_pin(dapm, "Main MIC");
 		}
 		ret = 0;
-		snd_kctl_jack_report(priv->snd_card, priv->headphone_kctl, 0);
+		snd_jack_report(jack->jack, 0);
 	}
 
 	return ret;
@@ -617,15 +617,19 @@ static int imx_wm8960_probe(struct platform_device *pdev)
 		priv->is_headset_jack = true;
 
 	if (gpio_is_valid(imx_hp_jack_gpio.gpio)) {
-		priv->headphone_kctl = snd_kctl_jack_new("Headphone", NULL);
-		ret = snd_ctl_add(priv->snd_card, priv->headphone_kctl);
-		if (ret)
-			dev_warn(&pdev->dev, "failed to create headphone jack kctl\n");
-
 		if (priv->is_headset_jack) {
 			imx_hp_jack_pin.mask |= SND_JACK_MICROPHONE;
 			imx_hp_jack_gpio.report |= SND_JACK_MICROPHONE;
 		}
+
+		priv->jack_data.card = data->card.snd_card;
+		INIT_LIST_HEAD(&priv->jack_data.kctl_list);
+
+		ret = snd_jack_add_new_kctl(&priv->jack_data,
+					"Headphone", imx_hp_jack_pin.mask);
+		if (ret)
+			dev_warn(&pdev->dev, "failed to create headphone jack kctl\n");
+
 		imx_hp_jack_gpio.jack_status_check = hp_jack_status_check;
 		imx_hp_jack_gpio.data = &imx_hp_jack;
 		ret = imx_wm8960_jack_init(&data->card, &imx_hp_jack,
-- 
1.7.9.5

